src/bgen/AttributeManager.cs

```diff
@@ -7,7 +7,15 @@
 using PlatformName = ObjCRuntime.PlatformName;
 #endif
 
-public class AttributeManager {
+public interface IAttributeManager {
```

Comment:
I'm not really a fan of using interfaces, because the experience with xharness has showed me that it becomes really annoying to navigate code, since "go to definition" on these method calls brings you to the interface, not the actual implementation.

For testing, I'd recommend a different approach: make these methods virtual, and override them in a mock instance.

```cs
public class AttributeManager {
    public virtual T[] GetCustomAttributes<T> (ICustomAttributeProvider provider) where T: Attribute {}
}
public class MockAttributeManager {
    public override T[] GetCustomAttributes<T> (ICustomAttributeProvider provider) where T: Attribute  {}
}
```

Reply:
Xharness was a diff story. The problem we had there is that we were bending over to match the expectations for an external project with a design that used DI. We had several implementations and we were trying to hide a lot of details. We can use virtual methods, moq does not care at all, but the issue with xharness has nothing to with interfaces, it has to do with DI. There is a reason why Miguel hated it, on the other had the aspnet core people love it.